src/QuadVelocityController.cpp

```diff
@@ -189,30 +223,120 @@ bool QuadVelocityController::waitUntilReady()
         return false;
     }
 
+    while (ros::ok()
+           && battery_msg_queue_.empty()
+           && ros::Time::now() < start_time + startup_timeout_) {
+        ros::spinOnce();
+        ros::Duration(0.005).sleep();
+    }
+
+    if (battery_msg_queue_.empty()) {
+        ROS_ERROR("Failed to fetch battery voltage");
+        return false;
+    }
+
     geometry_msgs::TransformStamped transform;
     bool success = getTransformAtTime(transform,
+                                      "quad",
+                                      "level_quad",
                                       ros::Time(0),
                                       startup_timeout_);
+    if (!success)
+    {
+        ROS_ERROR("Failed to fetch initial transform");
+        return false;
+    }
 
+    success = getTransformAtTime(transform,
+                                 "center_of_lift",
+                                 "level_quad",
+                                 ros::Time(0),
+                                 startup_timeout_);
     if (!success)
     {
         ROS_ERROR("Failed to fetch initial transform");
         return false;
     }
-    else
+
+    // Mark the last update time as 1ns later than the odometry message,
+    // because we should always have a message older than the last update
+    last_update_time_ = odometry_msg_queue_.front().header.stamp
+                      + ros::Duration(0, 1);
+    return true;
+}
+
+void QuadVelocityController::batteryCallback(const std_msgs::Float32& msg)
+{
+    ros::Time now = ros::Time::now();
+    if (!battery_msg_queue_.empty()
+          && now < battery_msg_queue_.back().header.stamp) {
+        ROS_ERROR_STREAM("Ignoring battery message at time "
+                      << now
+                      << " because the queue already has a message with stamp "
+                      << battery_msg_queue_.back().header.stamp);
+        return;
+    }
+
+    iarc7_msgs::Float64Stamped msg_to_queue;
+    msg_to_queue.data = msg.data;
+    msg_to_queue.header.stamp = now;
+    battery_msg_queue_.push_back(msg_to_queue);
+
+    // Keep only one message older than the last update time
+    std::vector<iarc7_msgs::Float64Stamped>::const_iterator first_geq_time
+        = std::lower_bound(battery_msg_queue_.begin(),
+                           battery_msg_queue_.end(),
+                           last_update_time_,
+                           &QuadVelocityController::timeVsMsgStampedComparator
+                                <iarc7_msgs::Float64Stamped>);
+    if (first_geq_time - 1 > battery_msg_queue_.begin()) {
+        battery_msg_queue_.erase(battery_msg_queue_.begin(),
+                                  first_geq_time - 1);
+    }
+}
+
+bool QuadVelocityController::getBatteryAtTime(
+        double& voltage,
+        const ros::Time& time,
+        const ros::Duration& timeout)
+{
+    // Wait until there's a message with stamp >= time
+    if (!waitForBatteryAtTime(time, timeout))
```

Comment:
Subtract a delay duration from here to allow less strict timing requirements for battery updates since they are sent much slower than transform updates?